core/arch/aarchxx/mangle.c

```diff
@@ -2922,7 +2954,573 @@ mangle_icache_op(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+    opnd_t op_res = opnd_create_reg(reg_to_pointer_sized(reg_res));
+    reg_id_t reg_st_val = opnd_get_reg(instr_get_src(instr, 0));
+    ASSERT(opnd_is_base_disp(instr_get_dst(instr, 0)) &&
+           opnd_get_index(instr_get_dst(instr, 0)) == DR_REG_NULL &&
+           opnd_get_disp(instr_get_dst(instr, 0)) == 0);
+    /* XXX i#4352: There is a bogus int opnd we have to account for here. */
+    bool is_pair = instr_num_srcs(instr) > 1 && opnd_is_reg(instr_get_src(instr, 1));
+    reg_t reg_base = opnd_get_base(instr_get_dst(instr, 0));
+    instr_t *no_match = INSTR_CREATE_label(dcontext);
+
+    /* The store could use 4 different regs (stxp), and we need 2 scratch regs to
+     * hold a pair of prior values, plus another for one of the new values (the other
+     * we put in the dead status reg): so 7 regs, but only 3 slots.  The stolen reg
+     * swap does not add because the store would then use one fewer of the scratch
+     * reg range.
+     */
+    reg_id_t scratch = DR_REG_NULL, scratch2 = DR_REG_NULL, scratch3 = DR_REG_NULL;
+    ushort slot, slot2, slot3;
+    bool should_restore = false, should_restore2 = false, should_restore3 = false;
+    if (!ldex_in_same_block) {
+        ASSERT(reg_orig_ld_val == DR_REG_NULL &&
+               reg_orig_ld_val2 == DR_REG_NULL); /* We only have to avoid swap_reg. */
+        /* We pass false to avoid the status reg, which we ourselves use. */
+        scratch = pick_scratch_reg(dcontext, instr, swap_reg, DR_REG_NULL, DR_REG_NULL,
+                                   false, &slot, &should_restore);
+        insert_save_to_tls_if_necessary(dcontext, ilist, instr, scratch, slot);
+#ifdef ARM
+        if (!is_cbnz_available(dcontext, reg_res)) {
+            /* We have no CBNZ so we need to preserve the flags. */
+            dr_save_arith_flags_to_reg(dcontext, ilist, instr, scratch);
+            PRE(ilist, instr,
+                instr_create_save_to_tls(dcontext, scratch, TLS_LDSTEX_FLAGS_SLOT));
+        }
+#endif
+        if (is_pair) {
+            scratch2 = pick_scratch_reg(dcontext, instr, swap_reg, scratch, DR_REG_NULL,
+                                        false, &slot2, &should_restore2);
+            scratch3 = pick_scratch_reg(dcontext, instr, swap_reg, scratch, scratch2,
+                                        false, &slot3, &should_restore3);
+            insert_save_to_tls_if_necessary(dcontext, ilist, instr, scratch2, slot2);
+            insert_save_to_tls_if_necessary(dcontext, ilist, instr, scratch3, slot3);
+        }
+        /* Compare address, arranging op_res to show failure on mismatch.
+         * XXX i#400: It is possible that the store could fault and the app could
+         * examine ops_res in the handler: i.e., it's not truly dead.  We do not
+         * account for that here.
+         */
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch, TLS_LDSTEX_ADDR_SLOT));
```

Comment:
If we failed to store the address from an actual prior excl-load, we would have an infinite loop, which would be relatively easy to detect and debug, so I'm not too worried there.  Plus, I don't think we could reasonably do any kind of check, since the app is free to have a bogus value, or execute an excl-store without a prior excl-load.